feat: [PON-49] Program subscription page design

mrazadar opened the pull request:
<!-- Anyone merging to this repository is expected to [promptly release and monitor their changes](https://openedx.atlassian.net/wiki/spaces/RS/pages/1835106870/How+to+contribute+to+our+repositories); if you are not able to do this DO NOT MERGE, please coordinate with someone who can to ensure that the changes are released.
-->
## Description

This PR introduces the new page design for Program subscription checkout experience. 
The changes includes: 

- Introduced a new route `/subscription` in `frontend-app-payment` MFE
- Reused existing payment functionality as much as possible 
- Introduced new components where necessary like `SubscriptionCartContent`, `SubscriptionOrderDetails`, `SubscriptionPaymentSubmitButton`, `SubscriptionLegal` etc
- Updated `package.json` to override the old `open-edx` theme with `@edx/brand-edx.org@^2.1.1`
- Responsive design changes 
- Refactored Alert List component
- Code cleanup for Subscription page (removing old unused code)

## Supporting information
https://2u-internal.atlassian.net/browse/PON-11
https://2u-internal.atlassian.net/browse/PON-49

## Testing instructions
This feature can't be tested until it get hooked with the new backend service for Program subscription experience. 

## Other information
<!-- 
Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, OpenEdx vs. edx.org differences, development vs. production environment differences, security, or accessibility.
-->
<!-- 
## Checklist
- [ ] Consider PCI compliance impact and whether this PR changes how credit card information is handled.
- [ ] Intend to [release and monitor](https://openedx.atlassian.net/wiki/spaces/RS/pages/1835106870/How+to+contribute+to+our+repositories) this PR promptly after merge.
-->
![Screen Shot 2023-03-01 at 5 41 02 PM](https://user-images.githubusercontent.com/5585262/222142321-8f2a993b-5308-4833-a6a6-b11811723032.png)
![Screen Shot 2023-03-01 at 5 41 14 PM](https://user-images.githubusercontent.com/5585262/222142334-c875798b-f632-4bbe-9384-4692fcdb2294.png)

mrazadar committed:
feat: PON-49 program subscription page design

feat: PON-49 new program subscription page

feat: PON-49 subscription design changes

test: updated failed snapshot tests

refactor: code cleanup, added form error styles

feat: code review minor changes

feat: PON-49 code improvements

feat: PON-49 reverted package.json changes

feat: PON-49 reverted package.json changes

julianajlk commented on src/payment/cart/TotalTable.jsx:
```diff
@@ -4,14 +4,22 @@ import { FormattedMessage } from '@edx/frontend-platform/i18n';
-        defaultMessage="TOTAL"
-        description="Label for the final total price of an order."
-      />
+      {!isSubscription ? (
```
Just a small preference to also follow `SummaryTable` component, swap the formatted messages so the ternary operator is `isSubscription ? ...`

mrazadar replied:
Done

julianajlk commented on src/payment/checkout/Checkout.jsx:
```diff
@@ -231,34 +231,35 @@ class Checkout extends React.Component {
+              {/* Apple Pay temporarily disabled per REV-927  - https://github.com/openedx/frontend-app-payment/pull/256 */}
+            </p>
+          </div>
+        ) : null}
```
Is there a benefit to using the ternary operator vs. `&&` here?  Is there a time when `isSubscription` could be undefined?

mrazadar replied:
No.
I know there will be no issue to use `&&` in our case but just following this practice when rendering [JSX conditionally](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx). Please let me know if you still want me to change it with `&&`

julianajlk replied:
Sounds good to keep as is.

julianajlk commented on src/payment/checkout/payment-form/StripePaymentForm.jsx:
```diff
@@ -145,12 +147,22 @@ const StripePaymentForm = ({
+          isProcessing={isProcessing}
+        />
+      ) : (
+        <PlaceOrderButton
```
Can you rename `PlaceOrderButton` component to `SubmitButton`? This will better match the subs button component name plus "place order" is the text in the button which could change in the future, so the name won't be relevant

mrazadar replied:
Will cover in the next subsequent PR against this [ticket](https://2u-internal.atlassian.net/browse/PON-82)

julianajlk commented on src/payment/checkout/payment-form/StripePaymentForm.jsx:
```diff
@@ -145,12 +147,22 @@ const StripePaymentForm = ({
-      />
+      {isSubscription ? (
+        <SubscriptionSubmitButton
+          onSubmitButtonClick={onSubmitButtonClick}
```
If you're planning to not have any of the functionality of this page done in this PR, can you comment out this onSubmit since it won't be that function?

mrazadar replied:
We eventually gonna change this handler I have added a TODO for now but left the `onSubmit` call there to test the error design state

julianajlk commented on src/subscription/SubscriptionPage.jsx:
```diff
@@ -0,0 +1,117 @@
+};
+
+const mapStateToProps = (state) => ({
+  ...paymentSelector(state),
```
`paymentSelector` will probably change if we don't fetch basket from ecommerce I am assuming - let's discuss in PL today, but if that's the case, if we're unsure, can you add a comment here for a future TODO if you don't plan to address this in this PR?

mrazadar replied:
We will try to update api, data layer against this [ticket](https://2u-internal.atlassian.net/browse/PON-82) in sprint 2. For now added a TODO here.

julianajlk replied:
Great, thank you!

julianajlk commented on src/subscription/cart/SubscriptionCart.jsx:
```diff
@@ -0,0 +1,134 @@
+            />
+          ))}
+
+          {isBulkOrder ? <UpdateQuantityForm /> : null}
```
I don't think we'll need any bulk order functionality in subs, so no need to add a quantity form (let's PL to double check)

julianajlk replied:
Updating: confirmed with Seth no bulk orders for subs. Feel free to remove anything that you've copied from the original payment flow that includes this on the subscription version.

mrazadar replied:
ok will remove while working on `pon-82`

julianajlk commented on src/subscription/cart/SubscriptionCart.jsx:
```diff
@@ -0,0 +1,134 @@
+            </OrderSummary>
+          ) : (
+            <>
+              <div className="skeleton py-2 mb-3 w-50" />
```
Not sure if this was copied from the regular Cart component, but for Subs we'll need less skeleton elements to reflect the simpler Summary section (I believe it would only have 1 line).

mrazadar replied:
We have new skeleton design which I will be updating against `PON-82`

julianajlk commented on src/subscription/cart/SubscriptionCart.jsx:
```diff
@@ -0,0 +1,134 @@
+            </>
+          )}
+        <SubscriptionOrderDetails />
+        <SubscriptionLegal price={summaryPrice} />
```
Do you think this could be a child of `SubscriptionOrderDetails` component instead of a sibling?

mrazadar replied:
Both can work.

julianajlk commented on src/subscription/cart/SubscriptionCart.jsx:
```diff
@@ -0,0 +1,134 @@
+        aria-relevant="all"
+        aria-label={intl.formatMessage(messages['payment.section.cart.label'])}
+      >
+        {loading ? <CartSkeleton /> : this.renderCart() }
```
We'll need to modify this `CartSkeleton` component to reflect the number of items subs will have (not just one course) or create a `SubscriptionCartSkeleton`. Imo, just adding 2 extra skeletons for the "In Your Cart" section if `isSubscription` is true, will suffice

mrazadar replied:
hmm.. TBH I'm not a big fan of content have loading skeletons. first they make the page slow and unresponsive, secondly they should represent page layout semantics instead of replicating the full application content. 

However, we can't be sure the exact number of products/courses associated with each program, so my suggestion is to keep what we have here.

julianajlk replied:
I haven't seen what UX has, but if what you have matches it than disregard my comment!

You're right that we can't know the exact number of courses, good call.

mrazadar replied:
I just noticed Michelle updated designs for loading states what wasn't available earlier so will be updating skeletons in upcoming PR.

julianajlk commented on src/subscription/cart/content/SubscriptionContent.jsx:
```diff
@@ -0,0 +1,67 @@
+        <h3 aria-level="2" className="mb-0">
+          <FormattedMessage
+            id="payment.subscription.purchase.cart.programTitle"
+            defaultMessage="Blockchain Fundamentals"
```
Touching on this point again, if this PR will not include functionality, can you add a TODO comment for hard coded temp default message here? But if you plan to add it in this PR later before merge, disregard all of my comments about this :)

mrazadar replied:
There is already a TODO added in the component documentation

julianajlk commented on src/subscription/cart/content/SubscriptionContent.jsx:
```diff
@@ -0,0 +1,67 @@
+      <p aria-level="2" className="body small mb-5">
+        <FormattedMessage
+          id="payment.subscription.purchase.cart.programOrganization"
+          defaultMessage="University of California, Berkeley"
```
Same as above.

mrazadar replied:
same as above.

julianajlk commented on src/subscription/cart/legal/SubscriptionLegal.jsx:
```diff
@@ -0,0 +1,41 @@
+}) => {
+  const supportLink = (
+    <Hyperlink
+      destination="https://support.edx.org/hc/en-us/sections/115004173027-Receive-and-Share-edX-Certificates"
```
Instead of hard coding the support URL here, can you get it from `getConfig()` instead and append the remainder? I believe it's `SUPPORT_URL`. It's best to get URLs from configs for open edX, but also in case they change in the future

mrazadar replied:
I will do it once we have actual URL from Michelle, However I have documented this in my follow up [ticket](https://2u-internal.atlassian.net/browse/PON-82).

julianajlk replied:
Great, thanks! We should remind her tomorrow about the URL.

mrazadar replied:
Resolving it for now already added as requirement in the next follow up ticket.

julianajlk commented on src/subscription/cart/legal/SubscriptionLegal.jsx:
```diff
@@ -0,0 +1,41 @@
+// import LocalizedPrice from '../../../payment/cart/LocalizedPrice';
+import messages from './messages';
+
+const SubscriptionLegal = ({
```
Is there a loading state and skeleton for this component?

mrazadar replied:
my guess is; we should not have loading states for static content. However if we really need to replicate the actual content with the page loading skeleton I can update the page skeleton in my next PR.

julianajlk replied:
Let's follow what UX has in the loading mock ups, which is one rectangle for this section.

julianajlk reviewed (changes requested):
Good work Raza! I added some comments/questions, let me know if you have questions.

Have you tested error handling?

Also, regarding a waffle flag in the backend to render this page, I don't think we'll need it because of the different route, however, let's discuss in stand up today if exposing this would be an issue.

julianajlk commented on src/payment/checkout/Checkout.jsx:
```diff
@@ -273,6 +274,7 @@ class Checkout extends React.Component {
               isBulkOrder={isBulkOrder}
```
This is a comment for L273, we'll need to add a new `handleSubmitStripeButtonClick` function for event tracking for subs, and pass that on as props, if the assumption is that we'll need a different event name than regular payments. Let's PL today.

mrazadar replied:
noted.

mrazadar commented:
Thank for the detailed reivew @julianajlk 
I have addressed the comments please review and LMK if there is a need of any further action on any item. 

Regarding tests: As discussed will implement them in future in a separate PR

julianajlk commented:
> Thank for the detailed reivew @julianajlk I have addressed the comments please review and LMK if there is a need of any further action on any item.
> 
> Regarding tests: As discussed will implement them in future in a separate PR

Thanks for the replies Raza! I resolved the ones that either don't need a reply from you OR if my response is just a 👍. I did leave some open threads for you to see my replies, even if they are not questions. Overall, I think we should look at the loading mocks UX has and maybe in the next PR you can share some screenshots of this state.

I forget that this is merging into the `ponderosa-master` branch and not `master`, so my question about exposing the new URL without functionality or using waffle flags from the BE won't be an issue until `ponderosa-master` is merged.

mrazadar commented:
> > Thank for the detailed reivew @julianajlk I have addressed the comments please review and LMK if there is a need of any further action on any item.
> > Regarding tests: As discussed will implement them in future in a separate PR
> 
> Thanks for the replies Raza! I resolved the ones that either don't need a reply from you OR if my response is just a 👍. I did leave some open threads for you to see my replies, even if they are not questions. There were some comments that can only be seen when you load all of them FYI.
> 
> Overall, I think we should look at the loading mocks UX has and maybe in the next PR you can share some screenshots of this state.
> 
> I forget that this is merging into the `ponderosa-master` branch and not `master`, so my question about exposing the new URL without functionality or using waffle flags from the BE won't be an issue until `ponderosa-master` is merged.


I have added replies to unresolved comments and resolved them. 

I just noticed Michelle updated designs for loading states that wasn't available earlier so will be updating skeletons in my upcoming PR.

shafqatfarhan commented on src/subscription/SubscriptionPage.jsx:
```diff
@@ -0,0 +1,118 @@
+      <div className="row subscription-page">
+        <h1 className="sr-only">
+          <FormattedMessage
+            id="payment.heading.page"
```
Please make this message ID more verbose

mrazadar replied:
Done

shafqatfarhan commented on src/subscription/cart/legal/SubscriptionLegal.jsx:
```diff
@@ -0,0 +1,41 @@
+import PropTypes from 'prop-types';
+import { Hyperlink } from '@edx/paragon';
+import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
+// import LocalizedPrice from '../../../payment/cart/LocalizedPrice';
```
Please remove this unused import

mrazadar replied:
done

shafqatfarhan commented on src/subscription/checkout/SubmitButton/SubscriptionSubmitButton.jsx:
```diff
@@ -0,0 +1,81 @@
+  showLoadingButton, onSubmitButtonClick, disabled, isProcessing, isSubmitted,
+}) => {
+  let submitButtonState = 'default';
+  // istanbul ignore if
+  if (disabled) { submitButtonState = 'disabled'; }
+  // istanbul ignore if
+  if (isProcessing) { submitButtonState = 'processing'; }
```
Please go through all the comments and remove them if they don't seem important

shafqatfarhan commented on src/subscription/index.scss:
```diff
@@ -0,0 +1,50 @@
+.subscription-page {
+  .heading-badge-wrapper {
+    display: flex;
+    justify-content: space-between;
+    align-items: flex-start;    
```
Use react-bootstrap style/classes here.

mrazadar replied:
Done

shafqatfarhan reviewed (approved):
Added my comments. Please remove changes made in the package.json file as edx brand theme is applied through webpack using the frontend-build package

mrazadar commented on src/payment/PaymentPage.jsx:
```diff
@@ -113,45 +105,12 @@ class PaymentPage extends React.Component {
-            'apple-pay-authorization-failure': intl.formatMessage(messages['payment.apple.pay.authorization.failure']),
-          }}
+        <FormattedAlertList
+          intl={this.props.intl}
```
used `useIntl` hook from react-intl here

mrazadar commented on src/payment/cart/SummaryTable.jsx:
```diff
@@ -13,17 +13,28 @@ const SummaryTable = ({ price }) => (
+      {
+        isSubscription ? (
+          <FormattedMessage
+            id="payment.summary.table.label.subscriptionPrice"
```
done

mrazadar commented:
> Added my comments. Please remove changes made in the package.json file as edx brand theme is applied through webpack using the frontend-build package

Removed
